deepsubtree.go

```diff
@@ -16,23 +17,134 @@ type DeepSubTree struct {
+		if err != nil {
+			return nil, err
+		}
+		dstNonExistenceProof.LeftSiblingProof, err = createExistenceProof(tree.ImmutableTree, leftSibling.key)
+		if err != nil {
+			return nil, err
+		}
+	}
+	if nonExistenceProof.Right != nil {
+		rightSibling, err := tree.getSiblingNode(nonExistenceProof.Right.Key)
+		if err != nil {
+			return nil, err
+		}
+		dstNonExistenceProof.RightSiblingProof, err = createExistenceProof(tree.ImmutableTree, rightSibling.key)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return &dstNonExistenceProof, nil
+}
+
+// Returns the sibling node of a leaf node with given key
+func (tree *ImmutableTree) getSiblingNode(key []byte) (*Node, error) {
+	siblingNode, err := tree.recursiveGetSiblingNode(tree.root, key)
+	if err != nil {
+		return nil, err
+	}
+	return siblingNode, nil
+}
+
+// Recursively iterates the tree to return the sibling node of a leaf node with given key
+func (tree *ImmutableTree) recursiveGetSiblingNode(node *Node, key []byte) (*Node, error) {
+	if node == nil || node.isLeaf() {
+		return nil, nil
+	}
+	leftNode, err := node.getLeftNode(tree)
+	if err != nil {
+		return nil, err
+	}
+	rightNode, err := node.getRightNode(tree)
+	if err != nil {
+		return nil, err
+	}
+	if leftNode != nil && leftNode.isLeaf() && bytes.Equal(leftNode.key, key) {
+		return rightNode, nil
+	}
+	if rightNode != nil && rightNode.isLeaf() && bytes.Equal(rightNode.key, key) {
+		return leftNode, nil
+	}
+
+	siblingNode, err := tree.recursiveGetSiblingNode(leftNode, key)
+	if err != nil {
+		return nil, err
+	}
+	if siblingNode != nil {
+		return siblingNode, nil
+	}
+	return tree.recursiveGetSiblingNode(rightNode, key)
+}
+
+func (node *Node) updateInnerNodeKey() {
+	if node.leftNode != nil {
+		node.key = node.leftNode.getHighestKey()
+	}
+	if node.rightNode != nil {
+		node.key = node.rightNode.getLowestKey()
+	}
+}
+
 // Traverses in the nodes in the NodeDB in the Deep Subtree
 // and links them together using the populated left and right
 // hashes and sets the root to be the node with the given rootHash
 func (dst *DeepSubTree) BuildTree(rootHash []byte) error {
-	if dst.root == nil {
-		rootNode, rootErr := dst.ndb.GetNode(rootHash)
-		if rootErr != nil {
-			return fmt.Errorf("could not set root of deep subtree: %w", rootErr)
-		}
-		dst.root = rootNode
-	} else if !bytes.Equal(dst.root.hash, rootHash) {
-		return fmt.Errorf(
-			"deep Subtree rootHash: %s does not match expected rootHash: %s",
-			dst.root.hash,
-			rootHash,
-		)
+	workingHash, err := dst.WorkingHash()
+	if err != nil {
+		return err
 	}
+	if !bytes.Equal(workingHash, rootHash) {
+		if dst.root == nil {
+			rootNode, rootErr := dst.ndb.GetNode(rootHash)
+			if rootErr != nil {
+				return fmt.Errorf("could not set root of deep subtree: %w", rootErr)
+			}
+			dst.root = rootNode
+		} else {
+			return fmt.Errorf(
+				"deep Subtree rootHash: %s does not match expected rootHash: %s",
+				workingHash,
+				rootHash,
+			)
+		}
```

Comment:
nit: I think inverting this a little is cleaner
```suggestion
		if dst.root != nil {
			return fmt.Errorf(
				"deep Subtree rootHash: %s does not match expected rootHash: %s",
				workingHash,
				rootHash,
			)
		}
		rootNode, rootErr := dst.ndb.GetNode(rootHash)
		if rootErr != nil {
			return fmt.Errorf("could not set root of deep subtree: %w", rootErr)
		}
		dst.root = rootNode
```